### PR TITLE
Add XLSForm Import

### PR DESCRIPTION
**Pull request by thenav56**

### CODE REVIEW

**Review by AdityaKhatri — APPROVED** [review]

File: src/components/questionnaire-builder/Questionnaire/index.tsx [hunk 1]
```diff
@@ -33,6 +33,7 @@ interface Props {
     onDelete: (id: number) => void;
     onEdit: (questionnaire: MiniQuestionnaireElement) => void;
     onXLSFormExport: (id: number) => void;
+    onkoboToolboxExport: (id: number) => void;
```
Comment (tnagorra) [inline src/components/questionnaire-builder/Questionnaire/index.tsx]: The function is not properly cased. Should be `onKoboToolboxExport`

File: src/typings/questionnaire.tsx [hunk 2]
```diff
@@ -89,6 +87,10 @@ export interface BaseQuestionElement {
     isRequired?: boolean;
 }
 
+export interface BaseQuestionElement extends BaseQuestionElementWithoutId {
```
Comment (tnagorra) [inline src/typings/questionnaire.tsx]: We do not need to create BaseQuestionElementWithoutId. We can use Omit utility type https://www.typescriptlang.org/docs/handbook/utility-types.html#omittk

File: src/views/Questionnaires/QuestionnaireList/index.tsx [hunk 3]
```diff
@@ -56,11 +59,39 @@ interface Params {
...
+    // Used by XLSForm Import
+    questionnaireCreateRequest: {
+        url: '/questionnaires/',
+        method: methods.POST,
+        body: ({ params: { body } = { body: undefined } }) => body,
+        onSuccess: ({
+            response,
+            params: { onSuccess } = {},
+        }) => {
+            if (onSuccess) onSuccess(response as QuestionnaireElement);
```
Comment (tnagorra) [inline src/views/Questionnaires/QuestionnaireList/index.tsx]: `if` body should always be in curly braces

File: src/views/Questionnaires/QuestionnaireList/index.tsx [hunk 4]
```diff
@@ -187,6 +223,132 @@ class QuestionnaireList extends React.PureComponent<Props, State> {
         this.props.requests.questionnaireRequest.do();
     }
 
+    private handleImportXLSForm = (files: File[]) => {
+        if (files.length <= 0) return;
+
+        const wb = new Excel.Workbook();
+        const reader = new FileReader();
+
+        const getColumnsIndex = ((columns: string[]) => (
```
Comment (tnagorra) [inline src/views/Questionnaires/QuestionnaireList/index.tsx]: We should use the helper function `listToMap`

File: src/views/Questionnaires/QuestionnaireList/index.tsx [hunk 5]
```diff
@@ -305,6 +466,11 @@ class QuestionnaireList extends React.PureComponent<Props, State> {
         });
     }
 
+    private handlekoboToolboxExport = (questionnaireId: number) => {
```
Comment (tnagorra) [inline src/views/Questionnaires/QuestionnaireList/index.tsx]: The casing should be `handleKoboToolboxExport`

File: src/views/Questionnaires/QuestionnaireList/index.tsx [hunk 6]
```diff
@@ -187,6 +223,132 @@ class QuestionnaireList extends React.PureComponent<Props, State> {
...
+                ...acc,
+                [col]: index,
+            }), {})
+        ));
+
+        reader.readAsArrayBuffer(files[0]);
+        reader.onload = () => {
+            const buffer = reader.result as ArrayBuffer;
+            wb.xlsx.load(buffer).then((workbook) => {
+                const survey = workbook.getWorksheet('survey');
```
Comment (tnagorra) [inline src/views/Questionnaires/QuestionnaireList/index.tsx]: Do we not need to verify if the worksheets are available?

File: src/views/Questionnaires/QuestionnaireList/index.tsx [hunk 7]
```diff
@@ -187,6 +223,132 @@ class QuestionnaireList extends React.PureComponent<Props, State> {
...
+                });
+
+                const body = {
+                    title: formTitle,
+                    project: this.props.projectId,
+                };
+
+                this.props.requests.questionnaireCreateRequest.do({
+                    body,
+                    onSuccess: (questionnaire: QuestionnaireElement) => (
```
Comment (tnagorra) [inline src/views/Questionnaires/QuestionnaireList/index.tsx]: The onSuccess method can be directly defined outside of the Component

**Pull request merged**